### PR TITLE
Check App ID

### PR DESCRIPTION
**Pull request by jeremad**

### CODE REVIEW

File: Identity/Crypto.cs [hunk 1]
```diff
@@ -40,6 +44,16 @@ internal static byte[] ObfuscateUserId(byte[] userId, byte[] trustchainId)
                 return GenericHash(toHash, BlockHashSize);
             }
 
+            internal static byte[] GenerateAppID(byte[] appSecret)
+            {
+                byte[] publicKey = new byte[PublicKeySize];
```
Comment (AlexandreBossard) [inline Identity/Crypto.cs]: You can use `var` here.

File: Identity/Crypto.cs [hunk 2]
```diff
@@ -74,6 +88,22 @@ public static bool VerifySignDetached(byte[] message, byte[] signature, byte[] p
...
+                    return false;
+                }
+                for (int i = 0; i < left.Length; i++)
+                {
+                    if (!left[i].Equals(right[i]))
+                    {
+                        return false;
+                    }
+                }
+                return true;
```
Comment (AlexandreBossard) [inline Identity/Crypto.cs:105]: There is a simpler approach (left.equal(right()). Which is slow, I know, but I don't think we should care. The proper way is to use the new `ReadOnlySpan<T>`, but we need some trickery to suppor every dotnet platforms. I would stick to the the boring left.Equal(right);

File: Identity/Crypto.cs [hunk 3]
```diff
@@ -40,6 +44,16 @@ internal static byte[] ObfuscateUserId(byte[] userId, byte[] trustchainId)
                 return GenericHash(toHash, BlockHashSize);
             }
 
+            internal static byte[] GenerateAppID(byte[] appSecret)
+            {
+                byte[] publicKey = new byte[PublicKeySize];
+                Array.Copy(appSecret, AppSecretSize - PublicKeySize, publicKey, 0, PublicKeySize);
+                byte[] start = new byte[1 + AuthoreSize];
```
Comment (AlexandreBossard) [inline Identity/Crypto.cs]: AuthorSize ?

File: Identity/Crypto.cs [hunk 4]
```diff
@@ -40,6 +44,16 @@ internal static byte[] ObfuscateUserId(byte[] userId, byte[] trustchainId)
...
+                start[0] = AppCreationNature;
+                var toHash = ConcatByteArrays(start, publicKey);
```
Comment (AlexandreBossard) [inline Identity/Crypto.cs:53]: There is `AuthorSize` byte of emptyness here ? Is it what you want ?

**Review by AlexandreBossard — CHANGES_REQUESTED** [review]

File: Identity/Crypto.cs (hunk 3 above)
Comment (jeremad) [reply Identity/Crypto.cs]: oopsy

File: Identity/Crypto.cs (hunk 4 above)
Comment (jeremad) [reply Identity/Crypto.cs:53]: yep, I need zeros here. Which is what c# does apparently

File: Identity/Crypto.cs (hunk 2 above)
Comment (jeremad) [reply Identity/Crypto.cs:105]: hum... `byte[]` does not implement `equal`, that's why I did this stupid implementation.
Or I missed something?

Comment (AlexandreBossard) [reply Identity/Crypto.cs:105]: https://docs.microsoft.com/en-us/dotnet/api/system.byte.equals?view=netcore-2.2

its `Equals()`, my bad.

**Review by AlexandreBossard — APPROVED** [review]

File: Identity/Crypto.cs (hunk 2 above)
Comment (jeremad) [reply Identity/Crypto.cs:105]: That's for a single byte struct, and that's what I did, right?

**Pull request merged**